### ISSUE TITLE
Method calls after opening connection: race condition

### ISSUE THREAD

**Issue by karlri** [diwic/dbus-rs#233]
Very useful crate, but I think I found a race condition. Perhaps I'm wrong, in that case greatful for advice!

The example code in this repo looks as follows:
```Rust
    let (resource, conn) = connection::new_session_sync()?;
    tokio::spawn(async {
        let err = resource.await;
```
The problem with this is that the MIO setup is done when resource is polled the first time. If a method call to DBUS is done right after spawning resource.await, then it may fail if unlucky with timing.

Right now this issue can be worked around by adding an async sleep after spawning resource.await.

This issue clould be avoided if connection::new_session_sync() set up FDs in mio before returning.

**Comment by diwic** [comment]
> The problem with this is that the MIO setup is done when resource is polled the first time. If a method call to DBUS is done right after spawning resource.await, then it may fail if unlucky with timing.

Interesting, but I'm not sure what the problem is here, i e, exactly what bad thing would happen if the timing is unlucky? 
Also, `dbus_connection_send` tries to send the message regardless of whether `mio` has set up the fds or not.

**Comment by karlri** [comment]
> Interesting, but I'm not sure what the problem is here, i e, exactly what bad thing would happen if the timing is unlucky?
It appears that the method call is sent, but there is no event to read the reply (I'm guessing because mio is set up after) and the call times out. I will see if I can construct a reproducible way of triggering the issue.

**Comment by karlri** [comment]
I was unable to construct a deterministic example that causes failure, but here follows one that fails after a few thousand iterations:
```Rust
use dbus_tokio::connection;
use dbus::nonblock;
use std::time::Duration;
use tokio::task;

#[tokio::main]
async fn main() {
    // Construct a local task set that can run `!Send` futures.
    let local = task::LocalSet::new();

    for i in 0..1000_000_000 {
        println!("{}", i);
        // Run the local task set.
        // This seems to increase the likelihood of failure on my machine.
        local.run_until(async move {
            // `spawn_local` ensures that the future is spawned on the local
            // task set.
            task::spawn_local(dbus_connect_then_call()).await.unwrap().unwrap();
        }).await;
    }
}



async fn dbus_connect_then_call() -> Result<(), Box<dyn std::error::Error>> {
    // Connect to the D-Bus session bus (this is blocking, unfortunately).
    let (resource, conn) = connection::new_session_sync()?;

    let calls = async {
        println!("Calling Hello...");
        let proxy = nonblock::Proxy::new("org.freedesktop.DBus", "/", Duration::from_secs(10), conn);
        // TODO: Handle timeouts and errors here
        let (names,): (Vec<String>,) = proxy.method_call("org.freedesktop.DBus", "ListNames", ()).await.unwrap();
    };

    tokio::select! {
        _ = calls => {},
        _ = resource => {}
    }

    Ok(())
}
```
Result:
```
3948
Calling Hello...
3949
Calling Hello...
3950
Calling Hello...
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: D-Bus error: Timeout waiting for reply (org.freedesktop.DBus.Error.Timeout)', src/libcore/result.rs:1165:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: JoinError::Panic(...)', src/libcore/result.rs:1165:5
```

**Comment by diwic** [comment]
So, even if mio sets up the fd after the first reply is received, there is still a call to `read_write` (and `process_all`) in the beginning of `poll_internal`. 

Thanks for the example! Do you mind if I copy it into the repository as a test/example?

I tried the following patch, and it seems to fix the race condition here. Could you see if it works for you as well?

    diff --git a/dbus-tokio/src/connection.rs b/dbus-tokio/src/connection.rs
    index ab407db..0c66225 100644
    --- a/dbus-tokio/src/connection.rs
    +++ b/dbus-tokio/src/connection.rs
    @@ -59,6 +59,7 @@ impl<C: AsRef<Channel> + Process> IOResource<C> {
         fn poll_internal(&mut self, ctx: &mut task::Context<'_>) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
             let c: &Channel = (*self.connection).as_ref();
     
    +        // println!("Poll_internal start");
             c.read_write(Some(Default::default())).map_err(|_| Error::new_failed("Read/write failed"))?;
             self.connection.process_all();
     
    @@ -76,12 +77,18 @@ impl<C: AsRef<Channel> + Process> IOResource<C> {
                 },
             };
     
    -        r.take_read_ready()?;
    -        r.take_write_ready()?;
    -
    -        if w.read { let _ = r.poll_read_ready(ctx)?; };
    -        if w.write { let _ = r.poll_write_ready(ctx)?; };
    +        if w.read {
    +            let _ = r.poll_read_ready(ctx)?;
    +        } else {
    +            r.take_read_ready()?;
    +        }
    +        if w.write {
    +            let _ = r.poll_write_ready(ctx)?;
    +        } else {
    +            r.take_write_ready()?;
    +        }
     
    +        // println!("Poll_internal end {:?}", w);
             Ok(())
         }
     }

**Comment by karlri** [comment]
Your patch above makes a lot of sense! Tested:

- MIPS 32 no fail after 100 000 iterations, then I aborted it, used to fail after ~3.
- ARM 32 no fail after 100 000 iterations, then I aborted it, used to fail after ~10.
- x86 64 no fail after 1 000 000 iterations, then I aborted it,  used to fail after ~5000.

The patch seems good, thanks a lot for the quick fix!

**Comment by karlri** [comment]
I looked at the docs for take_read_ready:

Doesn't that mean we cannot be awoken when events occur unless we ensure that poll_read_ready runs every call to poll_internal? How about this:

```
-        r.take_read_ready()?;
-        r.take_write_ready()?;
-
-        if w.read { let _ = r.poll_read_ready(ctx)?; };
-        if w.write { let _ = r.poll_write_ready(ctx)?; };
+        r.poll_read_ready(ctx)?;
+        r.poll_write_ready(ctx)?;
```

I don't fully understand the implications, but here goes some guesswork:
- ctx has a .waker() method, so poll_read_ready etc can request to re-run poll_internal when fd becomes readable.
- poll_read_ready needs to be run each call to poll_internal, because it has to re-register the wakeup each time? **I am very uncertain here.**
- if poll_internal is called when we can't read **(c.read == false)** meaning **no data available to read at the moment** then we don't call poll_read_ready(ctx) because of an if statement
- we can then end up in a situation where we cannot be awoken if readable

Sources:

take_read_ready: "This function is identical to poll_read_ready except that it will not notify the current task when a new event is received."
https://docs.rs/tokio/0.2.2/tokio/io/struct.Registration.html#method.take_read_read

"DBUS_WATCH_READABLE | As in POLLIN."
https://dbus.freedesktop.org/doc/api/html/group__DBusConnection.html#gga0556779e61aeb19eb9cf6b6466bd1b98a178bd7a88e9bfbb8505db76eb7e0ccc4

"POLLIN There is data to read."
http://man7.org/linux/man-pages/man2/poll.2.html

**Comment by diwic** [comment]
Hi,

So all this is a bit tricky, but if I understand things correctly:

The normal/idle state, when both in and out buffers are empty, is that watch.read = true and watch.write = false. 
watch.read = true means "poll me when data can be read" and watch.write = true means "poll me when data can be written".

So when you call `poll_write_ready`, that basically means "poll me when the out buffer is no longer full", which is essentially always, because the write buffer is always non-full (at least if we're not sending a huge message). 
So from just looking at your patch, it seems to be working because `poll_internal` is called over and over again - i e, the thread takes 100% CPU.

**Comment by karlri** [comment]
Hi! The way I interpreted the documentation for poll_write_ready, it will only wake up the task if:
1. out buffer is full
2. out buffer suddenly becomes no longer full => wake => poll_internal
Why? Because it says events are edge triggered:
https://docs.rs/tokio/0.2.11/tokio/io/struct.Registration.html#method.poll_write_ready
https://docs.rs/mio/0.6.21/mio/struct.Poll.html#edge-triggered-and-level-triggered
Perhaps I interpreted this incorrectly? I took down the PR. I tried adding a timeout (tokio::time::delay(..).await) after making the method call and receiving the reply it seems to use 0% cpu.

**Comment by diwic** [comment]
> keep implementation of poll_internal, share waker of IOResoruce with channel::Channel. always awake after method call to ffi::dbus_connection_send, should cover both signals and method calls?

Something like that. After `ffi::dbus_connection_send`, call `watch()`, and if `c.watch().write == true` then wake the task up.

**Comment by i-poper** [comment]
I'm sorry.
I accidentally pushed the issue to my forked repository with the Issue tag.
The diff is the exact content of the patch [previously posted](https://github.com/diwic/dbus-rs/issues/233#issuecomment-585038510) on this issue.

**Comment by diwic** [comment]
@karlri So I did some tests here as well, and it seems like your solution works best, although I can't explain fully why. So I have now pushed that.

Do you mind contributing your connect stresstest example to the dbus-rs repository?

**Comment by gh2o** [comment]
Using the basic scheduler causes this issue to happen 100% of the time, and the first method call always times out. The issue seems to have been fixed with #234 though. Awaiting the next dbus-tokio release. Thanks!

**Comment by diwic** [comment]
@gh2o Version 0.5.1 of dbus-tokio has now been released. Enjoy!
